src/serprog.c

```diff
@@ -0,0 +1,510 @@
+         * plus two CPU clock cycles. See Table 25-5 on page 240 for definition of
+         * minimum pulse width on RESET pin, t RST
+         * 2. Wait for at least 20 ms and then enable serial programming by sending
+         * the Programming Enable serial instruction to the SDO pin
+         * 3. The serial programming instructions will not work if the communication
+         * is out of synchronization. When in sync, the second byte (0x53) will echo
+         * back when issuing the third byte of the Programming Enable instruction
+         * ...
+         * If the 0x53 did not echo back, give RESET a positive pulse and issue a
+         * new Programming Enable command
+         */
+        unsigned char cs_mode = CS_MODE_DESELECTED;
+        if (perform_serprog_cmd(pgm, S_CMD_S_CS_MODE, &cs_mode, 1, NULL, 0) != 0) {
+            return -1;
+        }
+        usleep(5);
+        cs_mode = CS_MODE_SELECTED;
+        if (perform_serprog_cmd(pgm, S_CMD_S_CS_MODE, &cs_mode, 1, NULL, 0) != 0) {
+            return -1;
+        }
+        usleep(20000);
+
+        return -2;
+    }
+
+    return 0;
+}
+
+static int serprog_chip_erase(const PROGRAMMER *pgm, const AVRPART *p) {
+    unsigned char cmd[4], res[4];
+
+    if (!p->op[AVR_OP_CHIP_ERASE]) {
+        pmsg_error("chip erase instruction not defined for part %s\n", p->desc);
+        return -1;
+    }
+
+    memset(cmd, 0, sizeof(cmd));
+    avr_set_bits(p->op[AVR_OP_CHIP_ERASE], cmd);
+    pgm->cmd(pgm, cmd, res);
+    usleep(p->chip_erase_delay);
+    pgm->initialize(pgm, p);
+
+    return 0;
+}
+
+static void serprog_display(const PROGRAMMER *pgm, const char *p) {
+}
+
+static void serprog_enable(PROGRAMMER *pgm, const AVRPART *p) {
+}
+
+static void serprog_setup(PROGRAMMER *pgm) {
+    pgm->cookie = mmt_malloc(sizeof(struct pdata));
+}
+
+static void serprog_teardown(PROGRAMMER *pgm) {
+    mmt_free(pgm->cookie);
+    pgm->cookie = NULL;
+}
+
+static int serprog_parseextparams(const PROGRAMMER *pgm, const LISTID extparms) {
+    LNODEID ln;
+    const char *extended_param;
+    int rv = 0;
+
+    for (ln = lfirst(extparms); ln; ln = lnext(ln)) {
+        extended_param = ldata(ln);
+
+        if (str_starts(extended_param, "cs=")) {
+            unsigned int cs;
+            if (sscanf(extended_param, "cs=%u", &cs) != 1) {
+                pmsg_error("invalid chip select '%s'\n", extended_param);
+                rv = -1;
+            }
+            my.cs = cs;
+            continue;
+        }
+
+        if (str_eq(extended_param, "help")) {
+            msg_error("%s -c %s extended options:\n", progname, pgmid);
+            msg_error("  -xcs=cs_num    Sets the chip select (CS) to use on supported programmers.\n");
+            msg_error("                 Programmers supporting the 0x16 serprog command can have more than the default CS (0).\n");
+            msg_error("                 This option allows to choose this additional CS's (1,2,3,...) for programming the AVR.\n");
```

Comment:
`-xhelp` messages tend to be short. This level of detail is best left to the documentation (though you can leave it here too if you think it's important)